pallets/collator-allowlist/src/lib.rs

```diff
@@ -98,17 +101,25 @@ pub mod pallet {
 		/// Add the given `collator_id` to the allowlist.
 		/// Fails if
 		///   - `origin` fails the `ensure_root` check
+		///   - `collator_id` did not yet load their keys into the session pallet
 		///   - `collator_id` is already part of the allowlist
 		#[pallet::weight(<T as pallet::Config>::WeightInfo::allow())]
 		pub fn add(origin: OriginFor<T>, collator_id: T::ValidatorId) -> DispatchResult {
 			ensure_root(origin)?;
 
+			ensure!(
+				pallet_session::Pallet::<T>::is_registered(&collator_id),
+				Error::<T>::CollatorKeysNotLoaded
+			);
+
```

Comment:
Do we need to tightly-couple this pallet, with pallet-session? 
I would prefer to have an associated type that is bounded to `ValidatorRegistration` and wire the session pallet at runtime in.

I know, that this is mostly cosmetic at this point, but in my opinion using this loose-coupling hast two advantages and hence I would prefer to use this style as often as possible: 
* Testing can be mocked and does not depend on other pallets
* Refactoring is a lot easier in the future. 

Arguing against my case, that we specifically just want to check if the keys are loaded here, then I would prefer to not use the `ValidatorRegistration` interface but just check the storage directly as this makes it clear, what we want.

Reply:
@mustermeiszer both of your points are very good. So, first I did not appreciate designing this with such a thigh dependency style but ended up following the example I found at `pallet_anchors`, which seemed fair enough for this allowlist pallet given it's expected lifetime and given that it only makes sense - in practise - if used "on top" of the `pallet-session`.

> Testing can be mocked and does not depend on other pallets

Loosely coupling objects is indeed a major advantage when testing.

> Refactoring is a lot easier in the future.

As I pointed out, we don't expect a long lifetime for this pallet so I wouldn't put too much weight on that point in this case.

> Arguing against my case, that we specifically just want to check if the keys are loaded here, then I would prefer to not use the ValidatorRegistration interface but just check the storage directly as this makes it clear, what we want.

I appreciate the thought flexibility here. And I follow your sentiment that calling `Session::is_registered` is somewhat obscure for the intent at hand.

### Next steps

Given that it can be a good exercise for me to play with the loosely coupled pallet approach - which I just read through in the substrate docs `TIL` - I will give that a try and see what comes out. _If_ going that route introduces more indirection and less clarity than a tighter style, and given we only want this pallet for a short time period, I will keep it tight and just access storage directly to make the code more objective.

How does that sound?

Reply:
@mustermeiszer I have implemented the loose approach in https://github.com/centrifuge/centrifuge-chain/pull/535/commits/d90f7aec7693840c7f41b97d8e1dbc67736b55dd.

I like it. It's a trade-off and it was a good learning exercise for me, but overall I think this is good. I added a custom fn to deal with the unnatural language of "is_registered".

Let me know what you think, I am flexible.